[JIT] Introduce a fake Tensor creation node for IR unit tests

SplitInfinity opened the pull request:
**Summary**
There is often a need to create a Tensor when writing IR by hand for JIT
optimisation pass unit tests. The only options for this today are real
Tensor creation functions like `aten::ones`. Any test that uses these functions
must also use the same default arguments as the Python/C++ API, which means
that all of the tests have to be updated when the API is updated. This commit
introduces a new primitive, `prim::MakeTestTensor` with schema `() -> Tensor` that
should be used in unit tests instead of real Tensor creation functions. This new
primitive has no public-facing API, so the maintenance burden is much lower.

**Testing**
This commit updates the alias analysis and DCE tests to use `prim::MakeTestTensor` instead of
`aten::rand`, `aten::ones`, and `aten::zeros`.

```
$ ./bin/test_jit 
CUDA not available. Disabling CUDA and MultiCUDA tests
Note: Google Test filter = *-*_CUDA:*_MultiCUDA
[==========] Running 75 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 75 tests from JitTest
[ RUN      ] JitTest.ADFormulas
[       OK ] JitTest.ADFormulas (82 ms)
[ RUN      ] JitTest.Attributes
[       OK ] JitTest.Attributes (0 ms)
...
...
...
[ RUN      ] JitTest.LiteInterpreterPrim
[       OK ] JitTest.LiteInterpreterPrim (0 ms)
[ RUN      ] JitTest.LiteInterpreterLoadOrigJit
[       OK ] JitTest.LiteInterpreterLoadOrigJit (2 ms)
[----------] 75 tests from JitTest (150 ms total)

[----------] Global test environment tear-down
[==========] 75 tests from 1 test case ran. (150 ms total)
[  PASSED  ] 75 tests.
```

**Fixes**
This pull request fixes #33500.

Commit:
[JIT] Introduce a fake Tensor creation node for IR unit tests

Summary:
There is often a need to create a Tensor when writing IR by hand for JIT
optimisation pass unit tests. The only options for this today are real
Tensor creation functions like aten::ones. Any test that uses these functions
must also use the same default arguments as the Python/C++ API, which means
that all of the tests have to be updated when the API is updated. This commit
introduces a new primitive, prim::MakeTestTensor with schema () -> Tensor that
should be used in unit tests instead of real Tensor creation functions. This new
primitive has no public-facing API, so the maintenance burden is much lower.

Testing:
This commit updates the alias analysis and DCE tests to use prim::MakeTestTensor instead of
aten::rand, aten::ones, and aten::zeros.

SplitInfinity commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -746,6 +746,13 @@ RegisterOperators reg(
+           push(stack, at::Tensor());
+           return 0;
+         },
+         aliasAnalysisFromSchema()),
```
The tests fail if there is no registered operator. I guess that sort of makes sense, but it would be awesome if there was a way to register an operator just for the sake of the unit tests.

eellison replied:
You can register operators in test files. https://github.com/pytorch/pytorch/blob/806e7daa1fbe42380db04f7b44697975b3dd13c3/test/cpp/jit/test_misc.cpp#L954

suo replied:
Yeah, we should register whatever set of standard testing operators in a `test_base.cpp`.

SplitInfinity replied:
Added a `test_base.cpp` and registered the op there.

SplitInfinity commented:
`clang-tidy` CI job appears to have failed during setup. How do I rerun just that job?

suo commented:
> `clang-tidy` CI job appears to have failed during setup. How do I rerun just that job?

Looks like there's an upstream problem with the nvidia cuda repos. I'll bug them about it.

ezyang commented:
This is great! but I would recommend giving the prim tensor a name that makes it more obvious it's for testing only.

SplitInfinity commented:
Updates:
- Added `test_base.cpp` and moved the registration of the test operator there
- Renamed `prim::MakeTensor` to `prim::MakeTestTensor` to emphasize that it is for testing only

suo reviewed (approved):
Very nice!
